google-cloud-spanner/lib/google/cloud/spanner/interval.rb

```diff
@@ -0,0 +1,176 @@
+# Copyright 2017 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License")
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Google
+  module Cloud
+    module Spanner
+      class Interval
+        NANOSECONDS_IN_A_SECOND = 1000000000
+        NANOSECONDS_IN_A_MINUTE = NANOSECONDS_IN_A_SECOND * 60
+        NANOSECONDS_IN_AN_HOUR = NANOSECONDS_IN_A_MINUTE * 60
+        YEARS = 0
+        MONTHS = 1
+        DAYS = 2
+        HOURS = 4
+        MINUTES = 5
+        SECONDS = 6
+
+        def self.fromIso8601 interval_string
+          pattern = /
+            (?!$)(-?\d+Y)?(-?\d+M)?(-?\d+D)?(T(?=-?.?\d)(-?\d+H)?(-?\d+M)?(-?(((\d*)((\.|,)\d{1,9})?)|(\.\d{1,9}))S)?)?$
+            /
+          interval_months = 0
+          interval_days = 0
+          interval_nanoseconds = 0
+
+          captures = interval_string.match(pattern).captures
+          if captures.empty?
+            raise ArgumentError, "The ISO8601 provided was not in the correct format"
+          end
+
+          filter = Set[YEARS, MONTHS, DAYS, HOURS, MINUTES, SECONDS]
+          captures.each_with_index do |value, index|
+            unless filter.include? index
+              next
+            end
+
+            numeric_value = value.gsub(/[^0-9.,-]/, "")
```

Comment:
`String#to_i` will remove the trailing non-digit characters. So for the kinds of strings we have (numeric followed by zero or more non-numeric characters),

```ruby
numeric_value = value.gsub(/[^0-9.,-]/, "")
my_int = Integer numeric_value
```

is equivalent to the simpler and more idiomatic:

```ruby
my_int = value.to_i
```